### PR TITLE
#249 mpi

### LINKED ISSUE #249: #249 Manage multi processing with MPI following dyn-algo 114
follows [dynawo/dynawo-algorithms#116](https://github.com/dynawo/dynawo-algorithms/issues/114)

### PR DESCRIPTION
**Pull request by rosiereflo**
closes #249 
follows https://github.com/dynawo/dynawo-algorithms/pull/116

### CODE REVIEW

File: scripts/dynaflow-launcher.sh [hunk 1]
```diff
@@ -30,16 +30,55 @@ export LD_LIBRARY_PATH=$INSTALL/lib64:$INSTALL/lib:$LD_LIBRARY_PATH
...
+    if [ ! -z "$2" ]; then
+      USEMPI=true
+    fi
+    args="$args --contingencies"
+    shift
+    ;;
+  --nbThreads|-np)
+    NBPROCS=$2
+    shift # past argument
+    shift # past value
```
Comment (marcochiaramello) [inline scripts/dynaflow-launcher.sh:69]: shift 2 ?

File: scripts/envDFL.sh [hunk 2]
```diff
@@ -453,9 +490,26 @@ case $1 in
...
+
+# Nb Threads
+ARGS=""
+while (($#)); do
+    key="$1"
+    case $key in
+        --nbThreads|-np)
+        NBPROCS=$2
+        shift # past argument
+        shift # past value
```
Comment (marcochiaramello) [inline scripts/envDFL.sh:502]: shift 2 ?

**Review by marcochiaramello — CHANGES_REQUESTED** [review]

**Review by gautierbureau — COMMENTED** [review]
for me it is ok

File: scripts/dynaflow-launcher.sh (hunk 1 above)
Comment (rosiereflo) [reply scripts/dynaflow-launcher.sh]: solved

File: scripts/envDFL.sh (hunk 2 above)
Comment (rosiereflo) [reply scripts/envDFL.sh]: solved

File: README.md [hunk 3]
```diff
@@ -1,7 +1,8 @@
 # DynaFlow Launcher
 
 [![Build Status](https://github.com/dynawo/dynaflow-launcher/workflows/CI/badge.svg)](https://github.com/dynawo/dynaflow-launcher/actions)
-[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=dynawo_dynaflow-launcher&metric=alert_status)](https://sonarcloud.io/summary/new_code?id=dynawo_dynaflow-launcher)
+[![Quality Gate Status](https://sonarcloud.io/api/project_bEAD
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply README.md]: Nope! solved.

File: scripts/dynaflow-launcher.sh [hunk 4]
```diff
@@ -30,16 +30,55 @@ export LD_LIBRARY_PATH=$INSTALL/lib64:$INSTALL/lib:$LD_LIBRARY_PATH
 export IIDM_XML_XSD_PATH=$DYNAWO_HOME/share/iidm/xsd
 export DYNAFLOW_LAUNCHER_LOCALE=en_GB
 export DYNAFLOW_LAUNCHER_INSTALL=$INSTALL
-export DYNAWO_RESOURCES_DIR=$DYNAWO_HOME/share:$DYNAWO_HOME/share/xsd
-export DYNAWO_DDB_DIR=$DYNAWO_HOME/ddb
-export DYNAWO_IIDM_EXTENSION=$DYNAWO_HOME/lib/libdynawo_DataInterfaceIIDMExtension.so
-export DYNAWO_LIBIIDM_EXTENSIONS=$DYNAWO_HOME/lib
+export DYNAWO_RESOURCES_DIR=$DYNAWO_HOME/share:$DYNAWO_HOME/share/xsd:$DYNAFLOW_LAUNCHER_EXTERNAL_RESOURCES_DIR
+export DYNAWO_DDB_DIR=$DYNAFLOW_LAUNCHER_EXTERNAL_DDB $DYNAWO_HOME/ddb
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply scripts/dynaflow-launcher.sh]: You are right, only 1 path is supported. solved.

File: scripts/envDFL.sh [hunk 5]
```diff
@@ -371,47 +388,66 @@ build_tests_coverage() {
...
 
 launch() {
-    if [ ! -f $1 ]; then
-        error_exit "IIDM network file $network doesn't exist"
-    fi
     if [ ! -f $2 ]; then
-        error_exit "DFL configuration file $config doesn't exist"
+        error_exit "IIDM network file $1 doesn't exist"
+    fi
+    if [ ! -f $3 ]; then
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply scripts/envDFL.sh]: I think I got most of them.

**Review by gautierbureau — APPROVED** [review]

**Review by marcochiaramello — APPROVED** [review]

**Comment by gautierbureau** [comment]
```
RETURN_CODE=$?
return $RETURN_CODE
```

are missing in cmake_configure and cmake_build

File: 3rdParty/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+set(mpich_version "3.4.2")
+if(DEFINED ENV{DYNAWO_ALGORITHMS_MPICH_DOWNLOAD_URL})
+  set(mpich_prefix_url $ENV{DYNAWO_ALGORITHMS_MPICH_DOWNLOAD_URL})
+else()
+  set(mpich_prefix_url http://www.mpich.org/static/downloads)
+endif()
+set(mpich_url ${mpich_prefix_url}/${mpich_version}/mpich-${mpich_version}.tar.gz)
+set(mpich_md5 6ee1cfff98728e5160c6e78bdb1986ca)
+
+set(MPI_HOME ${CMAKE_INSTALL_PREFIX}/mpich)
```
Comment (gautierbureau) [inline 3rdParty/CMakeLists.txt:33]: remove to be able to find system mpich

File: 3rdParty/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+find_package(MPI QUIET)
+if(MPI_FOUND)
+  add_custom_target(mpich)
+  set(MPI_HOME ${CMAKE_INSTALL_PREFIX}/mpich)
```
Comment (gautierbureau) [inline 3rdParty/CMakeLists.txt:37]: remove to be able to find system mpich

**Review by gautierbureau — CHANGES_REQUESTED** [review]

File: 3rdParty/CMakeLists.txt (hunk 6 above)
Comment (rosiereflo) [reply 3rdParty/CMakeLists.txt]: solved

File: 3rdParty/CMakeLists.txt (hunk 7 above)
Comment (rosiereflo) [reply 3rdParty/CMakeLists.txt]: solved

**Review by gautierbureau — APPROVED** [review]

**Pull request merged**